casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala

```diff
@@ -78,6 +78,16 @@ object ProtoUtil {
       mainParent <- blocks.get(parentHash)
     } yield mainParent
 
+  def weightFromValidator(b: BlockMessage,
+                          validator: ByteString,
+                          blocks: collection.Map[ByteString, BlockMessage]): Int =
+    mainParent(blocks, b)
+      .map(weightMap(_).getOrElse(validator, 0))
+      .getOrElse(weightMap(b).getOrElse(validator, 0)) //no parents means genesis -- use itself
```

Comment:
We'll have to double check our assumptions at the end (especially ordering). Like are we guaranteed at this point to have an otherwise valid block - because it could be an invalid block with no parents.

Reply:
That's why I check separately if the block is genesis and anything else that returns 0 is invalid. https://github.com/rchain/rchain/pull/940/files#diff-eddb46712698d75072b3f453346dd0afR70

Reply:
Ok then being pedantic this comment should be `// no parents means genesis or invalid` as when this function is called from blockSender it is never the genesis

Reply:
Eh, seems like you merged. Don't bother if it is a pain to push.